[Tools] bitcoin-wallet - a tool for creating and managing wallets offline

jnewbery opened the pull request:
Adds an offline tool `bitcoin-wallet-tool` for wallet creation and maintenance.

Currently this tool can create a new wallet file and display information on an existing wallet. It can later be extended to support other common wallet maintenance tasks (eg run the salvage and zapwallettxes maintenance tasks on an existing wallet).

Doing wallet maintenance tasks in an offline tool makes much more sense (and is potentially safer) than having to spin up a full node.

jonasschnelli committed:
[tools] Add wallet inspection and modification tool

This commit adds wallet-tool, a tool for creating and interacting with
wallet files. Original implementation was by Jonas Schnelli
<dev@jonasschnelli.ch> with modifications by John Newbery
<john@johnnewbery.com>

MSVC files were provided by Chun Kuan Lee <ken2812221@gmail.com>:

build: Add MSVC project files for bitcoin-wallet-tool

jnewbery commented:
This is currently a work in progress. It is a rebase of @jonasschnelli's #8745. Please see that PR for history and context. I've volunteered to maintain this through to merge. Thanks Jonas for your work in writing and maintaining this branch so far!

It is currently rebased on top of #12490, which removes a circular dependency and allows it to build. Only the _Add wallet inspection and modification tool "bitcoin-wallet-tool"_ commit is relevant here.

#12490 can't be merged until after the V0.17 branch is forked, so detailed review at this point would be immature. Concept and high-level comments most definitely welcomed.

domob1812 commented:
Concept ACK, looks useful.  Will that tool also support things like creating addresses or dumping privkeys in the future, or "just" actual maintenance of the wallet file?  That's something that might be useful for offline systems (although you can of course spin up the daemon with an empty data dir).

jnewbery commented:
> Will that tool also support things like creating addresses or dumping privkeys in the future

Potentially, yes. That would be a discussion for a future PR though :)

domob1812 commented:
@jnewbery:  Yes, of course - let's go in steps.  For me, those functions would have been useful from time to time, though, so I'd love to see that added in some follow up.  (And am happy to work on it myself if that helps.)

MeshCollider commented:
Concept ACK, does seem clean and useful to have these as a separate tool.

laanwj commented:
Thanks for picking this up, I agree an external wallet manipulation/recovery tool is useful

jnewbery commented:
Thanks all concept ACKers! If you want to help with this PR, please review #12490, since that's a pre-req to get this to build. That PR just removes deprecated code, so it should be an easy review.

practicalswift commented on src/bitcoin-wallet-tool.cpp:
```diff
@@ -0,0 +1,106 @@
+
+static void SetupWalletToolArgs()
+{
+    std::string usage;
```
Unused variable :-)

practicalswift commented on src/wallet/wallettool.h:
```diff
@@ -0,0 +1,19 @@
+#include <wallet/wallet.h>
+
+namespace WalletTool {
+CWallet* CreateWallet(const std::string name, const fs::path& path);
```
`name` should be passed by const reference?

practicalswift commented on src/wallet/wallettool.h:
```diff
@@ -0,0 +1,19 @@
+
+namespace WalletTool {
+CWallet* CreateWallet(const std::string name, const fs::path& path);
+CWallet* LoadWallet(const std::string name, const fs::path& path);
```
`name` should be passed by const reference?

jnewbery commented:
#12490 has been merged. I've rebased on master and removed the [WIP] tag.

This is ready for review.

MarcoFalke commented on src/bitcoin-wallet-tool-res.rc:
```diff
@@ -0,0 +1,35 @@
+        BLOCK "040904E4" // U.S. English - multilingual (hex)
+        BEGIN
+            VALUE "CompanyName",        "Bitcoin"
+            VALUE "FileDescription",    "bitcoin-wallet-tool (CLI Bitcoin wallet editor utility)"
```
nit: replace `Bitcoin` with `"PACKAGE_NAME "`, since it doesn't work on other Bitcoin wallets.

jnewbery replied:
agree. Changed

MarcoFalke commented on src/bitcoin-wallet-tool.cpp:
```diff
@@ -0,0 +1,105 @@
+    RandomInit();
+    try {
+        if(!WalletAppInit(argc, argv))
+        return EXIT_FAILURE;
```
nit: This indentation is misleading. Should probably use the clang-format-diff script to fix all of these.

jnewbery replied:
Thanks. I've used clang-format-diff to fix these.

maflcko reviewed (commented):
haven't looked at the code yet. Just two style nits.

jnewbery commented:
Rebased on master and fixed @MarcoFalke's comments.

Sjors commented:
How does this compare to #10102 (@ryanofsky) which also creates an independent wallet executable that could, with some tweaks, be used offline? Obviously this change is much simpler.

jnewbery commented:
> How does this compare to #10102

Very different. 10102 is much more ambitious and is a change to the overall architecture of bitcoind. It has dependency on #10973, so I don't expect it to be merged any time soon. It also doesn't currently support running the wallet executable separately as far as I'm aware.

This is a small, self-contained tool for created, inspecting and updating wallet files.

promag reviewed (commented):
Concept ACK. Are you planning adding tests? In this PR or other?

jnewbery commented:
> Are you planning adding tests? In this PR or other?

Yes, I plan to add tests to this PR.

practicalswift commented on src/bitcoin-wallet-tool.cpp:
```diff
@@ -0,0 +1,103 @@
+        return EXIT_FAILURE;
+    }
+    if (argc < 2 || HelpRequested(gArgs)) {
+        std::string usage = strprintf(_("%s wallet-tool version"), PACKAGE_NAME) + " " + FormatFullVersion() + "\n" +
```
Should be `;` instead of ` +` at the end of the line?

practicalswift commented on src/bitcoin-wallet-tool.cpp:
```diff
@@ -0,0 +1,103 @@
+    if (!WalletTool::executeWalletToolFunc(method, name))
+        return EXIT_FAILURE;
+    ECC_Stop();
+    return true;
```
```
bitcoin-wallet-tool.cpp:102:5: warning: bool literal returned from 'main' [-Wmain]
    return true;
    ^      ~~~~
```

practicalswift commented on src/wallet/wallettool.cpp:
```diff
@@ -0,0 +1,134 @@
+{
+    if (fs::exists(path)) {
+        fprintf(stderr, "Error: File exists already\n");
+        return NULL;
```
```
src/wallet/wallettool.cpp:17:16: warning: use nullptr [modernize-use-nullptr]
```

Please replace `NULL` with `nullptr` throughout :-)

practicalswift commented on src/wallet/wallettool.cpp:
```diff
@@ -0,0 +1,134 @@
+    fprintf(stdout, "Wallet info\n===========\n");
+    fprintf(stdout, "Encrypted: %s\n", wallet_instance->IsCrypted() ? "yes" : "no");
+    fprintf(stdout, "HD (hd seed available): %s\n", wallet_instance->GetHDChain().seed_id.IsNull() ? "no" : "yes");
+    fprintf(stdout, "Keypool Size: %lu\n", (unsigned long)wallet_instance->GetKeyPoolSize());
```
```
src/wallet/wallettool.cpp:81:45: warning: consider replacing 'unsigned long' with 'uint64' [google-runtime-int]
```

practicalswift replied:
Why the cast?

jnewbery commented:
Thanks for the review @promag and @practicalswift . All comments addressed in the fixup commit. I'll squash if you're happy with the changes.

jnewbery commented on src/wallet/wallettool.cpp:
```diff
@@ -0,0 +1,134 @@
+        }
+        fprintf(stdout, "Zaptxs successful executed. Please rescan your wallet.\n");
+    } else {
+        fprintf(stderr, "Unknown command\n");
```
done

jonasschnelli commented:
Concept ACK
Thanks for picking this up...

jnewbery commented:
Thanks @promag . I've fixed the segfault and squashed everything into one commit.

jnewbery commented:
@practicalswift - I've fixed the CWallet leaks in f81d8c8ad4718e2158434bf330962a08d9219b61. Please review - I'll squash if you're happy.

MarcoFalke commented on src/Makefile.am:
```diff
@@ -505,6 +517,34 @@ bitcoin_tx_LDADD = \
+  $(LIBMEMENV) \
+  $(LIBSECP256K1)
+
+bitcoin_wallet_tool_LDADD += $(BOOST_LIBS) $(BDB_LIBS) $(SSL_LIBS) $(CRYPTO_LIBS) $(MINIUPNPC_LIBS) $(EVENT_PTHREADS_LIBS) $(EVENT_LIBS) $(ZMQ_LIBS)
```
No need to attempt to link `SSL_LIBS`. If the tool wanted to upload the private keys, might as well do it without TLS

MarcoFalke replied:
Probably the same for `$(MINIUPNPC_LIBS) $(EVENT_PTHREADS_LIBS) $(EVENT_LIBS) $(ZMQ_LIBS)`?

jnewbery replied:
Thanks Marco. Fixed.

practicalswift commented:
@jnewbery Leaks gone :-)

jnewbery commented:
Addressed https://github.com/bitcoin/bitcoin/pull/13926#discussion_r217413912 , squashed and pushed.

promag commented:
@jnewbery 
```
➜  bitcoin git:(wallet_tool) ✗ ./src/bitcoin-wallet-tool -name=/tmp/w1 create
Topping up keypool...
Wallet info
===========
Encrypted: no
HD (hd seed available): yes
Keypool Size: 2000
Transactions: 0
Address Book: 0
➜  bitcoin git:(wallet_tool) ✗ ./src/bitcoin-wallet-tool -name=/tmp/w1 create
Error: File exists already
➜  bitcoin git:(wallet_tool) ✗ ./src/bitcoin-wallet-tool -name=/tmp/w1/foo/.. create
Topping up keypool...
Wallet info
===========
Encrypted: no
HD (hd seed available): yes
Keypool Size: 2000
Transactions: 0
Address Book: 0
```
the 3rd run should fail right?

promag commented on src/wallet/wallettool.cpp:
```diff
@@ -0,0 +1,145 @@
+    delete wallet;
+}
+
+static std::shared_ptr<CWallet> CreateWallet(const std::string& name, const fs::path& path)
```
These could return `std::unique_ptr`?

jnewbery replied:
Indeed. Fixed

jnewbery replied:
Actually, I'll keep these as shared

promag replied:
Why?

promag commented on src/wallet/wallettool.h:
```diff
@@ -0,0 +1,20 @@
+std::shared_ptr<CWallet> CreateWallet(const std::string& name, const fs::path& path);
+std::shared_ptr<CWallet> LoadWallet(const std::string& name, const fs::path& path);
+void WalletShowInfo(CWallet* wallet_instance);
+bool executeWalletToolFunc(const std::string& method, const std::string& file);
```
nit, `bool ExecuteWalletToolFunc...`.

jnewbery replied:
fixed

promag reviewed (commented):
While running
```
./src/bitcoind -wallet=/tmp/w1 -regtest
```
This happens:
```
./src/bitcoin-wallet-tool -name=/tmp/w1 info
libc++abi.dylib: terminating with uncaught exception of type std::runtime_error: BerkeleyBatch: Failed to open database environment.
[1]    78930 abort      ./src/bitcoin-wallet-tool -name=/tmp/w1 info
```

MarcoFalke commented on src/Makefile.am:
```diff
@@ -505,6 +517,34 @@ bitcoin_tx_LDADD = \
+  $(LIBBITCOIN_UTIL) \
+  $(LIBBITCOIN_CRYPTO) \
+  $(LIBUNIVALUE) \
+  $(LIBBITCOIN_ZMQ) \
```
Could also remove these:

```diff
diff --git a/src/Makefile.am b/src/Makefile.am
index 41cd066fa0..fc398a5fd9 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -539,8 +539,6 @@ bitcoin_wallet_tool_LDADD = \
   $(LIBBITCOIN_CONSENSUS) \
   $(LIBBITCOIN_UTIL) \
   $(LIBBITCOIN_CRYPTO) \
-  $(LIBUNIVALUE) \
-  $(LIBBITCOIN_ZMQ) \
   $(LIBLEVELDB) \
   $(LIBLEVELDB_SSE42) \
   $(LIBMEMENV) \

jnewbery replied:
removed

MarcoFalke commented on src/Makefile.am:
```diff
@@ -505,6 +517,34 @@ bitcoin_tx_LDADD = \
+  $(LIBMEMENV) \
+  $(LIBSECP256K1)
+
+bitcoin_wallet_tool_LDADD += $(BOOST_LIBS) $(BDB_LIBS) $(CRYPTO_LIBS)
```
Might need `$(MINIUPNPC_LIBS)` to pass travis?

jnewbery replied:
added

MarcoFalke commented:
@promag Would be better to write these up as python tests?

promag commented:
@MarcoFalke ok, I'll push a commit with the above tests and for the remaining commands.

ryanofsky commented:
I think it'd be nice to call the executable `bitcoin-wallet` instead of `bitcoin-wallet-tool`. I seems unusual to give a command an extra suffix like `-tool`, though I guess `bitcoin-cli` also has a similarly descriptive suffix.

jnewbery commented:
> I think it'd be nice to call the executable bitcoin-wallet instead of bitcoin-wallet-tool

Seems reasonable to me. We already have `bitcoin-tx` and not `bitcoin-tx-tool`. I'll change the name to `bitcoin-wallet`.

promag committed:
[tests] Add wallet-tool test

Original tests by João Barbosa <joao.paulo.barbosa@gmail.com>

Additional contribution by John Newbery <john@johnnewbery.com>

promag commented:
@jnewbery should disallow multiple `-name`, like `-name=/tmp/w1  -name=/tmp/w2 info`?

jnewbery commented on src/wallet/wallettool.cpp:
```diff
@@ -0,0 +1,145 @@
+    fprintf(stdout, "Address Book: %zu\n", wallet_instance->mapAddressBook.size());
+}
+
+bool executeWalletToolFunc(const std::string& method, const std::string& name)
```
changed

jnewbery commented on src/wallet/wallettool.cpp:
```diff
@@ -0,0 +1,145 @@
+        }
+        fprintf(stdout, "Zaptxs successful executed. Please rescan your wallet.\n");
+    } else {
+        fprintf(stderr, "Unknown command\n");
```
yep. Better

jnewbery commented:
> While running
>
> [...]
>
> This happens:
>
> [...]

I now catch the exception in LoadWallet in wallettool.cpp

jnewbery commented:
Rebased on master and fixed all review comments so far.

ryanofsky reviewed (commented):
Reviewed like 75% of this, but wanted to post comments so far.

promag commented:
nit, add src/bitcoin-wallet to .gitignore.

Sjors reviewed (commented):
Concept ACK.

This tool could also be useful as a migration tool if we ever decide to move away from DBD and deprecate that dependency from bitcoind/qt. Already it could have its own `--with-incompatible-bdb` configure argument (can wait for a future PR).

@theuni should probably take a look at the build system aspects of this.

Needs testing in Windows

How do I use this with `-testnet`? Do we want to ignore `bitcoin.conf`. If so, that needs mentioning in the help text. Longer term it seems better to support that, e.g. once creating new addresses is possible, `addresstype` should be honored.

I assume fresh wallets trigger a rescan because the tool doesn't know the block height? If so, maybe warn about that, and add a `heighthint` argument later.

promag commented on src/bitcoin-wallet.cpp:
```diff
@@ -0,0 +1,103 @@
+static void SetupWalletToolArgs()
+{
+    gArgs.AddArg("-?", "This help message", false, OptionsCategory::OPTIONS);
+    gArgs.AddArg("-name=<wallet-name>", "Specify wallet name", false, OptionsCategory::OPTIONS);
```
Also have mixed feeling here. IMO it should be consistent with `bitcoind` to begin with. For instance, with `psql` (postgres client) ou can set the database in multiple ways: env var, `-d` option or positional argument.

MarcoFalke commented:
Mind to cherry-pick the tests from  #14283, so that they can be reviewed in one go with the current code?

promag commented:
@MarcoFalke if the test strategy looks good then I should clean it before.

jonasschnelli commented:
Tested a bit...
I think the help should mention that the created wallet is relative to the Bitcoin data dir (from reading the help, I would have expected that the wallet will be created in the current directory).

**I personally would prefer that the tool does not interact with the datadir.** (specify path or expect current dir).

Also, how can you create a new wallet in the regtest or testnet subdirectory?

ryanofsky commented:
> I personally would prefer that the tool does not interact with the datadir. 

I think that would be a sensible idea for a low-level wallet debugging tool, but a bad idea for a tool that can replace `bitcoind` and `bitcoin-qt` as a way of accessing wallets. I think the least confusing way for bitcoin-wallet to work is for it to use the same configuration `bitcoind` and `bitcoin-qt` do, and to open and locate wallets in exactly the same way (with exact same `-wallet` `-walletdir` `-datadir` `-regtest` `-conf` etc parameters).

Lets think about an actual use case. If a user who runs `bitcoind` or `bitcoin-qt` normally encounters a problem that requires them to run "salvage" or "zaptxs", it would be nice if they could just run `bitcoin-wallet salvage` `bitcoin-wallet zaptxs` from the command line directly, without having to `cd` into weird directories or specify long data paths or learn an new set of command line arguments different from the existing ones.

The `bitcoin-wallet` tool is also going to be extended in the future. I think it'd be nice if it had `dump` and `load` subcommands that could be a safe for backing up and transferring wallets into flat files without the user having to know anything about the internal structure of their data dir. I think it'd be nice if there were a `shell` subcommand that would drop into an rpcconsole like shell and allow running wallets rpc commands in an offline mode.

ken2812221 commented on test/functional/tool_wallet.py:
```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env python3
```
Add this file into test_runner.py script list?

jnewbery replied:
done

jonasschnelli commented:
> I think that would be a sensible idea for a low-level wallet debugging tool, but a bad idea for a tool that can replace bitcoind and bitcoin-qt as a way of accessing wallets.

I'm not sure about this but I can follow your thoughts. I think one can also argue that the wallet-tool is a low-level accessing tool for experts.

> Lets think about an actual use case. If a user who runs bitcoind or bitcoin-qt normally encounters a problem that requires them to run "salvage" or "zaptxs".

If you have to salvage or zap transactions, you are in low-level debug mode IMO.


However, I think we should...
a) ... interpret the provided wallet name as filename (with optional path components)
b) ... if wallet has not been found, check the default datadir/walletdir if it contains the wallet
c) ... allow -regtest/-testnet switch
d) ... describe the path behaviour in the help-text

ryanofsky commented:
What you're suggesting sounds good, as long as it doesn't mean (part (a) is vague) that `bitcoin-wallet` should interpret the `-wallet=<path>` option differently from `bitcoind` and `bitcoin-qt`

The `salvage` and `zaptxs` subcommands are somewhat low level, but I don't think a `dump` command would be so obscure, and I think it would be nice if different various tools understood the same options so for example:

```sh
bitcoin-qt -testnet -wallet=foo
bitcoin-wallet -testnet -wallet=foo dump > foo.txt
```

could work without the without the user having to figure out where their datadir is located, where the testnet directory lives inside datadir, whether the testnet directory contains a "wallets" subdirectory, or is from a legacy install that preceded it, and so on.

jnewbery commented:
Rebased on master. tool_wallet.py is failing.

Sjors commented:
What happens if you run this tool on a wallet that's currently opened by `bitcoind`?

`zapwallettxes` is [sometimes](https://www.reddit.com/r/Bitcoin/comments/59h8f4/how_to_delete_unconfirmed_transactions_on_bitcoin/) recommended to deal with stuck transactions. Especially before the GUI had basic RBF support, but might still be useful (though it's a pretty blunt instrument).

practicalswift commented on src/wallet/wallettool.cpp:
```diff
@@ -0,0 +1,152 @@
+            fprintf(stderr, "Salvage failed\n");
+            return false;
+        }
+        //TODO, set wallets best block to genesis to enforce a rescan
```
nit: Should have a space between `//` and comment.

jnewbery replied:
fixed

ken2812221 commented on src/bitcoin-wallet.cpp:
```diff
@@ -0,0 +1,103 @@
+
+int main(int argc, char* argv[])
+{
+    SetupEnvironment();
```
Could you follow #13883 to add a Windows argument handler here?

jnewbery replied:
done

jnewbery commented on test/functional/tool_wallet.py:
```diff
@@ -0,0 +1,101 @@
+# Copyright (c) 2018 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test bitcoint-wallet-tool.
```
fixed

jnewbery commented on test/functional/tool_wallet.py:
```diff
@@ -0,0 +1,101 @@
+
+        # TODO calling info on a wallet already opened results in the segfault:
+        #  `libc++abi.dylib: terminating with uncaught exception of type std::runtime_error: BerkeleyBatch: Failed to open database environment`
+        #  the error must be fixed and tested here
```
This is now fixed, and the test case has been added to `tool_wallet.py`

jnewbery commented:
I've updated the tool to treat `-datadir`, `-wallet`, `-regtest` and `-testnet` arguments in the same way that bitcoind treats those arguments. I've also added help text to explain where it looks for wallet files:

```
wallet-tool is an offline tool for creating and interacting with Bitcoin Core wallet files.
By default wallet-tool will act on wallets in the default wallet directory in the datadir.
To change the target wallet, use the -datadir, -wallet and -testnet/-regtest arguments.
```

I believe all comments so far are now addressed.

ken2812221 commented:
@jnewbery the filelock on Windows is broken now, so the test would fail. See #14426

jnewbery commented:
New push:

- Fix the build failure with --disable-wallet
- skip the test if wallet is not compiled
- Catch when the wallet file can't be opened earlier
- rebased on #14426 to try to fix the windows failure

This PR now depends on #14426 .

ryanofsky reviewed (approved):
utACK 719ae605df48fe4dc4d1329b83cb0069160d40dc. There are some things I would want to clean up here (mainly deduping code and sharing more functionality between `bitcoind` and `bitcoind-wallet`), but this seems good as a starting point for a new tool that can be extended later.

jnewbery commented:
rebased

ryanofsky reviewed (approved):
utACK ebd8417a6b8fbb1575140f0f04b568f17a47914a. No changes since last review other than rebasing after windows file lock #14426 merge.

Sjors commented on src/Makefile.am:
```diff
@@ -509,6 +521,32 @@ bitcoin_tx_LDADD = \
 bitcoin_tx_LDADD += $(BOOST_LIBS) $(CRYPTO_LIBS)
 #
 
+# bitcoin-wallet-tool binary #
```
Nag nit (though don't bother if there's no other changes needed)

jnewbery replied:
fixed!

Sjors commented on src/bitcoin-wallet.cpp:
```diff
@@ -0,0 +1,117 @@
+    if (argc < 2 || HelpRequested(gArgs)) {
+        std::string usage = strprintf("%s bitcoin-wallet version", PACKAGE_NAME) + " " + FormatFullVersion() + "\n\n" +
+                                      "wallet-tool is an offline tool for creating and interacting with Bitcoin Core wallet files.\n" +
+                                      "By default wallet-tool will act on wallets in the default wallet directory in the datadir.\n" +
```
Nit: "the datadir for mainnet" (since `testnet=1` in `bitcoin.conf` is ignored)

jnewbery replied:
fixed

Sjors reviewed (approved):
My earlier [comment](
https://github.com/bitcoin/bitcoin/pull/13926#discussion_r220178473) about `-name` vs. `-wallet` is addressed now. I still think `./bitcoin-wallet create blah` is more intuitive than `./bitcoin-wallet -wallet=blah create`, but can live with both.

Trying to load a wallet that's currently opened by bitcoind now nicely returns a `Error loading . Is wallet being used by other process?`

Lightly tested ebd8417 on macOS 10.14. If you call `bitcoin-wallet -testnet` without arguments it crashes with `Segmentation fault: 11` instead of showing the help.

MarcoFalke commented on configure.ac:
```diff
@@ -1111,7 +1118,7 @@ else
   AC_CHECK_LIB([ssl],         [main],SSL_LIBS=-lssl, AC_MSG_ERROR(libssl missing))
 
-  if test x$build_bitcoin_cli$build_bitcoin_tx$build_bitcoind$bitcoin_enable_qt$use_tests != xnonononono; then
+  if test x$build_bitcoin_wallet$build_bitcoin_cli$build_bitcoin_tx$build_bitcoind$bitcoin_enable_qt$use_tests != xnononononono; then
```
I think you no longer link against libevent, so this could be removed?

jnewbery replied:
removed

practicalswift commented on src/wallet/wallettool.cpp:
```diff
@@ -0,0 +1,157 @@
+    }
+
+    std::shared_ptr<CWallet> wallet_instance(new CWallet(name, WalletDatabase::Create(path)), WalletToolReleaseWallet);
+    bool first_run;
```
Reduce scope of `first_run` :-)

jnewbery replied:
fixed :)

MeshCollider commented:
I think this will be a really useful tool to add more functionality too that isn't really suited to adding to bitcoind itself. Just testing this (have done some light code review, didn't look at the `build_msvc/` files or `configure.ac` & `Makefile.am` changes).  A few early comments:

Some of @ryanofsky's comments mentioned above still need addressing, e.g. that the `-printtoconsole` and `-debug` args still aren't recognized and error if they are tried to be used. I am slightly worried about the code duplication too but I think its ok.

Calling `info` on a non-existent wallet should throw and error, but instead it creates a file and prints info of a wallet containing no keys, transactions, etc. Then calling create on the same wallet name will error that the file already exists.

One other note, the tool should check a command was provided. Currently it just exits without doing anything if no command is provided.

MeshCollider commented on src/wallet/wallettool.cpp:
```diff
@@ -0,0 +1,157 @@
+    } else if (command == "info") {
+        std::string error;
+        if (!WalletBatch::VerifyEnvironment(path, error)) {
+            fprintf(stderr, "Error loading %s. Is wallet being used by other process?\n", name.c_str());
```
nit: "another process"

jnewbery replied:
fixed

jnewbery commented:
Thanks for testing @MeshCollider. I intend to pick this up again next week and address all feedback.

practicalswift commented:
I'm afraid this PR doesn't compile when rebased on `master` :-)

practicalswift commented on src/bitcoin-wallet.cpp:
```diff
@@ -0,0 +1,117 @@
+static bool WalletAppInit(int argc, char* argv[])
+{
+    SetupWalletToolArgs();
+    std::string error;
```
Could use variable name `errorMessage` instead to avoid shadowing `error` from `src/util/system.h`?

jnewbery replied:
fixed

jnewbery commented:
Rebased on master: https://github.com/bitcoin/bitcoin/compare/ebd8417a6b8fbb1575140f0f04b568f17a47914a..bd8875b50fd2d774cb98263450c0ad150aa374d7
Fixed build issues from rebase: https://github.com/bitcoin/bitcoin/compare/bd8875b50fd2d774cb98263450c0ad150aa374d7..21b4b35a8ecf879ec5909820ab36b3aa29358fe7

jnewbery commented on src/wallet/wallettool.cpp:
```diff
@@ -0,0 +1,152 @@
+    return wallet_instance;
+}
+
+static std::shared_ptr<CWallet> LoadWallet(const std::string& name, const fs::path& path)
```
Yes. I agree that this could be refactored to remove duplicate code. Can we save that for a follow-up PR?

jnewbery commented on src/bitcoin-wallet.cpp:
```diff
@@ -0,0 +1,103 @@
+        argc--;
+        argv++;
+    }
+    std::vector<std::string> args = std::vector<std::string>(&argv[1], &argv[argc]);
```
Changed to not construct a vector at all.

jnewbery commented:
Addressing @MeshCollider's feedback:

> Some of @ryanofsky's comments mentioned above still need addressing, e.g. that the -printtoconsole and -debug args still aren't recognized and error if they are tried to be used. I am slightly worried about the code duplication too but I think its ok.

These should all be addressed now.

> Calling info or zaptxs on a non-existent wallet should throw and error, but instead it creates a file and prints info of a wallet containing no keys, transactions, etc. Then calling create on the same wallet name will error that the file already exists.

This is fixed.

> One other note, the tool should check a command was provided. Currently it just exits without doing anything if no command is provided.

This is fixed.

jnewbery commented:
Latest force-push should address all remaining feedback: https://github.com/bitcoin/bitcoin/compare/21b4b35a8ecf879ec5909820ab36b3aa29358fe7..d05ec61f7659466e911ea3ae48aa583c792d8849

Thanks everyone for review! Please let me know if I've missed anything.

jnewbery commented:
rebased

ryanofsky reviewed (approved):
utACK 7064a9515dd7de8d32a1e4e7c08cb0ce17677b23. Changes since last review: rebase, tweaks like adding better error checking, supporting debug/printtoconsole.

Deduplicating wallet load & create wallet code between bitcoind & bitcoin-wallet would be nice cleanup to follow up with in the future.

Sjors commented:
I find the following behavior a bit scary:

```sh
$ src/bitcoin-wallet help zaptxs
Zaptxs successful executed. Please rescan your wallet.
```

I understand that we don't want to duplicate the `help` in this initial version, but it should probably throw an error.

I'm also confused how `zaptxs` is supposed to work here. In `bitcoind` `-zapwallettxes` takes an argument and actually leads to a rescan. When I call `bitcoin-wallet zaptxs` and then open the zapped wallet the transactions and labels are still there.

I'm not sure how to test `salvage`.

If `bitcoin.conf` sets `testnet=1` there's no way to acces a mainnet wallet (not a big deal).

The wallet creation command doesn't take arguments (like watch-only). I suppose that's fine for this initial version.

ryanofsky commented:
I think it would be good to merge wallet tool and make a new issue or set of issues to keep track of the usability problems and feature deficits Sjors is bringing up.

I don't think we should be trying to perfect this tool in a single github PR. I think it would be better to merge it, mark it as experimental, and continue to develop it in git. This would make any proposed changes easier to understand and discuss.

ryanofsky commented:
> The wallet creation command doesn't take arguments (like watch-only). I suppose that's fine for this initial version.

One idea for this might be to remove most functionality from the wallet tool, and just allow it to fill `JSONRPCRequest` objects and call the limited subset of RPC methods that work offline. This would also let it share named/positional argument parsing code with bitcoin-cli.

Sjors commented:
Either marking as experimental or reducing feature set works for me, though I think the latter is a safer approach.

jnewbery commented:
Thanks for the input @Sjors and @ryanofsky .

> I find the following behavior a bit scary:
> ...

Good catch! I've added that as an error condition (with test) in https://github.com/bitcoin/bitcoin/pull/13926/commits/c904860a24aade9a59f9948eb5e8e3da80a7b12f

> I don't think we should be trying to perfect this tool in a single github PR

Yes. Good point. I've removed the `salvage` and `zaptxs` commands in this PR, so this tool now can only provide info or create a wallet. I'll move the `salvage` and `zaptxs` commands to a follow-up PR. Done in https://github.com/bitcoin/bitcoin/pull/13926/commits/3519b2f0fb0f8fa22a788471d32917cefad61d82.

jnewbery commented:
Squashed fixup commits: https://github.com/bitcoin/bitcoin/compare/3519b2f0fb0f8fa22a788471d32917cefad61d82..a92b6085ea1c371bde447d96f22f6f17d3357348

jgarzik commented:
@ryanofsky 's suggestion matches the original intent when the new tools (e.g. `bitcoin-tx`) were introduced, and plans were made for others (`bitcoin-script`, `bitcoin-key`, and `bitcoin-wallet`):

* Start small.
* Get it in-tree.
* Discuss and grow from there.

Once of Linus's favorite wisdoms was that the Internet is the best test lab in the world, and your goal should be to accelerate changes to the point where they are being tested in the field by real users.  Wider audience = more testing and real world feedback.

Small+shipping is a good risk-adjusting model that maximizes the amount of field testing and review.   And the common pattern in OSS of endlessly nitpicking small details winds up being counter-productive to higher code quality (by restricting testing to The People Who Follow This Unmerged PR).  

Getting software into the field also validates basic developer assumptions of the software's value, utility and design.

Sjors commented:
One more thing: `src/bitcoin-wallet` needs to be added to `.ignore`.

Otherwise tACK a92b608

jnewbery commented:
> src/bitcoin-wallet needs to be added to .gitignore.

Done: https://github.com/bitcoin/bitcoin/compare/a92b6085ea1c371bde447d96f22f6f17d3357348..2aeac46d17dc74d488494c77b9d1852faeb7dc16

Sjors commented:
re-tACK 2aeac46

MarcoFalke commented on src/bitcoin-wallet.cpp:
```diff
@@ -0,0 +1,121 @@
+
+    if (!fs::is_directory(GetDataDir(false))) {
+        fprintf(stderr, "Error: Specified data directory \"%s\" does not exist.\n", gArgs.GetArg("-datadir", "").c_str());
+        return EXIT_FAILURE;
```
This is not defined behavior, since EXIT_FAILURE has no defined type nor value, but you need a bool(false). c.f. https://en.cppreference.com/w/c/program/EXIT_status

jnewbery replied:
I don't understand this. There are plenty of other locations where `return EXIT_FAILURE` is used (and even a couple of PRs where `return false` has been changed to `return EXIT_FAILURE` - #9067 and #13349)

MarcoFalke replied:
We are not in `main` here, which has a return type `int`. The function signature here is `static bool WalletAppInit`, so the return type is `bool`. Commonly `EXIT_FAILURE` is `1`, so `bool(1) == true`, which off-by-one.

You can also test this trivially by providing a datadir that does not exist. It will print an error, but continue and print garbage.

jnewbery replied:
Duh. Thanks! I only looked at the github snippet. Will fix.

MarcoFalke commented:
Concept ACK. Going to test and merge this this week. Should be safe, since the only methods are `info` and `create`

MarcoFalke commented on src/wallet/wallettool.cpp:
```diff
@@ -0,0 +1,139 @@
+static void WalletToolReleaseWallet(CWallet* wallet)
+{
+    wallet->WalletLogPrintf("Releasing wallet\n");
+    wallet->Flush();
```
question: The `Flush()` calls in this file accept a bool that should be set to true on shutdown?

jnewbery replied:
Yes, you're right. This got missed out when rebasing on a commit that changed the function signature for `Wallet::Flush()`.

I think the same is true for https://github.com/bitcoin/bitcoin/blob/d8794a78a887a920276c7124f1c46d69592c6c4e/src/wallet/wallet.cpp#L96

Fixed: #15390
